File: API/1001-1100/1094_Multi_Indicator_Swing/PY/multi_indicator_swing_strategy.py
import clr

clr.AddReference("StockSharp.Messages")
clr.AddReference("StockSharp.Algo")
clr.AddReference("StockSharp.Algo.Indicators")
clr.AddReference("StockSharp.Algo.Strategies")

from System import TimeSpan
from StockSharp.Messages import DataType, CandleStates
from StockSharp.Algo.Indicators import SimpleMovingAverage, RelativeStrengthIndex
from StockSharp.Algo.Strategies import Strategy


class multi_indicator_swing_strategy(Strategy):
    def __init__(self):
        super(multi_indicator_swing_strategy, self).__init__()
        self._candle_type = self.Param("CandleType", DataType.TimeFrame(TimeSpan.FromHours(1))) \
            .SetDisplay("Candle Type", "Type of candles", "General")
        self._sma_length = self.Param("SmaLength", 20) \
            .SetGreaterThanZero() \
            .SetDisplay("SMA Length", "SMA period", "Indicators")
        self._rsi_length = self.Param("RsiLength", 14) \
            .SetGreaterThanZero() \
            .SetDisplay("RSI Length", "RSI period", "Indicators")
        self._rsi_oversold = self.Param("RsiOversold", 45.0) \
            .SetDisplay("RSI Oversold", "RSI oversold", "Indicators")
        self._rsi_overbought = self.Param("RsiOverbought", 55.0) \
            .SetDisplay("RSI Overbought", "RSI overbought", "Indicators")
        self._signal_cooldown_bars = self.Param("SignalCooldownBars", 12) \
            .SetGreaterThanZero() \
            .SetDisplay("Signal Cooldown", "Bars to wait between new entries", "Trading")
        self._prev_close = 0.0
        self._prev_sma = 0.0
        self._prev_rsi = 0.0
        self._has_prev = False
        self._cooldown_remaining = 0

    @property
    def candle_type(self):
        return self._candle_type.Value

    @candle_type.setter
    def candle_type(self, value):
        self._candle_type.Value = value

    def OnReseted(self):
        super(multi_indicator_swing_strategy, self).OnReseted()
        self._prev_close = 0.0
        self._prev_sma = 0.0
        self._prev_rsi = 0.0
        self._has_prev = False
        self._cooldown_remaining = 0

    def OnStarted2(self, time):
        super(multi_indicator_swing_strategy, self).OnStarted2(time)
        self._prev_close = 0.0
        self._prev_sma = 0.0
        self._prev_rsi = 0.0
        self._has_prev = False
        self._cooldown_remaining = 0
        self._sma = SimpleMovingAverage()
        self._sma.Length = self._sma_length.Value
        self._rsi = RelativeStrengthIndex()
        self._rsi.Length = self._rsi_length.Value
        subscription = self.SubscribeCandles(self.candle_type)
        subscription.Bind(self._sma, self._rsi, self.OnProcess).Start()

    def OnProcess(self, candle, sma_val, rsi_val):
        if candle.State != CandleStates.Finished:
            return
        sv = float(sma_val)
        rv = float(rsi_val)
        close = float(candle.ClosePrice)
        if self._cooldown_remaining > 0:
            self._cooldown_remaining -= 1
        if self._has_prev:
            os_level = float(self._rsi_oversold.Value)
            ob_level = float(self._rsi_overbought.Value)
            trend_up = close > sv and sv > self._prev_sma
            trend_down = close < sv and sv < self._prev_sma
            cross_above_sma = self._prev_close <= self._prev_sma and close > sv
            cross_below_sma = self._prev_close >= self._prev_sma and close < sv
            long_entry = trend_up and cross_above_sma and self._prev_rsi <= os_level and rv > os_level
            short_entry = trend_down and cross_below_sma and self._prev_rsi >= ob_level and rv < ob_level
            if self._cooldown_remaining == 0:
                if long_entry and self.Position <= 0:
                    self.BuyMarket()
                    self._cooldown_remaining = self._signal_cooldown_bars.Value
                elif short_entry and self.Position >= 0:
                    self.SellMarket()
                    self._cooldown_remaining = self._signal_cooldown_bars.Value
        self._prev_close = close
        self._prev_sma = sv
        self._prev_rsi = rv
        self._has_prev = True

    def CreateClone(self):
        return multi_indicator_swing_strategy()

File: API/1001-1100/1094_Multi_Indicator_Swing/CS/MultiIndicatorSwingStrategy.cs
using System;
using System.Linq;
using System.Collections.Generic;

using Ecng.Common;
using Ecng.Collections;
using Ecng.Serialization;

using StockSharp.Algo.Indicators;
using StockSharp.Algo.Strategies;
using StockSharp.BusinessEntities;
using StockSharp.Messages;

namespace StockSharp.Samples.Strategies;

/// <summary>
/// Multi indicator swing strategy using RSI and SMA for direction.
/// </summary>
public class MultiIndicatorSwingStrategy : Strategy
{
	private readonly StrategyParam<DataType> _candleType;
	private readonly StrategyParam<int> _smaLength;
	private readonly StrategyParam<int> _rsiLength;
	private readonly StrategyParam<decimal> _rsiOversold;
	private readonly StrategyParam<decimal> _rsiOverbought;
	private readonly StrategyParam<int> _signalCooldownBars;

	private decimal? _prevClose;
	private decimal? _prevSma;
	private decimal? _prevRsi;
	private int _cooldownRemaining;

	public DataType CandleType { get => _candleType.Value; set => _candleType.Value = value; }
	public int SmaLength { get => _smaLength.Value; set => _smaLength.Value = value; }
	public int RsiLength { get => _rsiLength.Value; set => _rsiLength.Value = value; }
	public decimal RsiOversold { get => _rsiOversold.Value; set => _rsiOversold.Value = value; }
	public decimal RsiOverbought { get => _rsiOverbought.Value; set => _rsiOverbought.Value = value; }
	public int SignalCooldownBars { get => _signalCooldownBars.Value; set => _signalCooldownBars.Value = value; }

	public MultiIndicatorSwingStrategy()
	{
		_candleType = Param(nameof(CandleType), TimeSpan.FromHours(1).TimeFrame())
			.SetDisplay("Candle Type", "Type of candles", "General");
		_smaLength = Param(nameof(SmaLength), 20)
			.SetGreaterThanZero()
			.SetDisplay("SMA Length", "SMA period", "Indicators");
		_rsiLength = Param(nameof(RsiLength), 14)
			.SetGreaterThanZero()
			.SetDisplay("RSI Length", "RSI period", "Indicators");
		_rsiOversold = Param(nameof(RsiOversold), 45m)
			.SetDisplay("RSI Oversold", "RSI oversold", "Indicators");
		_rsiOverbought = Param(nameof(RsiOverbought), 55m)
			.SetDisplay("RSI Overbought", "RSI overbought", "Indicators");
		_signalCooldownBars = Param(nameof(SignalCooldownBars), 12)
			.SetGreaterThanZero()
			.SetDisplay("Signal Cooldown", "Bars to wait between new entries", "Trading");
	}

	/// <inheritdoc />
	protected override void OnReseted()
	{
		base.OnReseted();

		_prevClose = null;
		_prevSma = null;
		_prevRsi = null;
		_cooldownRemaining = 0;
	}

	/// <inheritdoc />
	protected override void OnStarted2(DateTime time)
	{
		base.OnStarted2(time);

		var sma = new SMA { Length = SmaLength };
		var rsi = new RelativeStrengthIndex { Length = RsiLength };

		var subscription = SubscribeCandles(CandleType);

		subscription
			.Bind(sma, rsi, (candle, smaVal, rsiVal) =>
			{
				if (candle.State != CandleStates.Finished)
					return;

				if (!IsFormedAndOnlineAndAllowTrading())
					return;

				if (_cooldownRemaining > 0)
					_cooldownRemaining--;

				var previousClose = _prevClose;
				var previousSma = _prevSma;
				var previousRsi = _prevRsi;

				if (previousClose.HasValue && previousSma.HasValue && previousRsi.HasValue)
				{
					var trendUp = candle.ClosePrice > smaVal && smaVal > previousSma.Value;
					var trendDown = candle.ClosePrice < smaVal && smaVal < previousSma.Value;
					var crossAboveSma = previousClose.Value <= previousSma.Value && candle.ClosePrice > smaVal;
					var crossBelowSma = previousClose.Value >= previousSma.Value && candle.ClosePrice < smaVal;
					var longEntry = trendUp && crossAboveSma && previousRsi.Value <= RsiOversold && rsiVal > RsiOversold;
					var shortEntry = trendDown && crossBelowSma && previousRsi.Value >= RsiOverbought && rsiVal < RsiOverbought;

					if (_cooldownRemaining == 0)
					{
						if (longEntry && Position <= 0)
						{
							var volume = Volume + Math.Abs(Position);
							BuyMarket(volume);
							_cooldownRemaining = SignalCooldownBars;
						}
						else if (shortEntry && Position >= 0)
						{
							var volume = Volume + Math.Abs(Position);
								SellMarket(volume);
							_cooldownRemaining = SignalCooldownBars;
						}
					}
				}

				_prevClose = candle.ClosePrice;
				_prevSma = smaVal;
				_prevRsi = rsiVal;
			})
			.Start();

		var area = CreateChartArea();
		if (area != null)
		{
			DrawCandles(area, subscription);
			DrawIndicator(area, sma);
			DrawOwnTrades(area);
		}
	}
}
